### ISSUE TITLE
Support Aggregation for Bucketing on Value Change

### ISSUE THREAD

**Issue by gingerwizard** [elastic/elasticsearch#14267]
A recent requirement to solve the gaps and island problem using Elasticsearch.

https://www.simple-talk.com/sql/t-sql-programming/the-sql-of-gaps-and-islands-in-sequences

Currently, this is being solved with an inefficient aggregation script.

This ticket proposes we support the ability to create buckets when the value for a field changes.  This assumes data is sorted appropriately prior to the aggregation.

For example, consider the following data.  The documents are sorted by the "doc_id" field. We create buckets on changes on the "value" field.

{ "index": { "_index": "test", "_type": "test", "_id": "1" } }
{ "value": "A"}
{ "index": { "_index": "test", "_type": "test", "_id": "2" } }
{ "value": "A" }
{ "index": { "_index": "test", "_type": "test", "_id": "3" } }
{ "value": B"}
{ "index": { "_index": "test", "_type": "test", "_id": "4" } }
{ "value": "A"}
{ "index": { "_index": "test", "_type": "test", "_id": "5" } }
{ "value": "A" }

Here we end up with 3 buckets. "A" - 2, "B" -1, "A" - 2

This would require us to label buckets with appropriately distinct names - possibly with the sequence range.
The sort field above could easily be a timestamp and the field to detect change on a numeric.

_What constitutes a change and thus new bucket?_

I think this largely depends on type.  If a string field is used, the change should be discrete values i.e. any change in value.  For other types we could support different models with a "model" parameter. For starters,  numerics could bucket based on the differential - the user would be required to specify the desired interval as a rate of change e.g. if they expect a continous gap > 0, 0 if they expect values to not change.  
Like any aggregation metrics/sub aggregations could be applied as desired.

Proposed syntax:

```
GET /test/_search
{
  "query": {
    "match_all": {}
  },
  "aggs": {
    "islands": {
      "change": {
        "model":"linear",
        "rate":0,
        "field":"value"
      }
    }
  }
}

```

Thanks
Dale

**Comment by gingerwizard** [comment]
@markharwood @polyfractal Please comment.

**Comment by gingerwizard** [comment]
Correction: The proposal is to do this as a pipeline aggregation.  The above syntax above is therefore incorrect. The issue discussing @HonzaKral is that to make this useful its unlikely people will want to "sample" buckets and will want to inspect each value - and thus lots of buckets, which will be expensive.

**Comment by markharwood** [comment]
Trying to understand the motivation more. 
Is it possible that these pre-collapsed buckets would have child aggs? Would this bucket-collapsing exercise be expected to fuse their results?

**Comment by polyfractal** [comment]
Hmm, I think I misunderstood the problem originally.  I didn't realize it needed to be sorted in the context of one field (e.g. an `id`) but checking for gaps in another field.  I think that is going to make it tricky to do as a regular aggregation.  

IIRC, aggregations can only evaluate a single ValueSource currently, so they can only inspect the values of one field.  I'm also not sure how sorting plays into it, since I think the sorting is done *while* documents are collecting, which is also when aggregations are being collected.  So the agg won't see any sorted values, just a random stream of documents.

If this was for a island/gapping on a single, numeric field (e.g. on the ID itself) it would be less of a problem because ordering isn't important, just tracking numerical gaps.

> Correction: The proposal is to do this as a pipeline aggregation. 

I don't think it makes sense to implement this as a pipeline.  Pipeline aggs only operate on the results of buckets produced by other aggs, which means all the documents have long since been lost to the query.  So you'd be finding islands/gaps in bucketed data, which I don't think makes sense?

**Comment by yaccob** [comment]
I think that the **sorting criteria needs to be evident** to enable sorting the resulting buckets. I'm trying to demonstrate this with a practical use case. 

## Simple Use Case

Below you can see a use case that demonstrates why awareness of the sorting criteria may be important.
In this use case the user wants to see an aggregated view of his transactions per account and the **chronological balance development** across accounts. Since this use case is presenting a chronological view, the time field is the sorting criteria for this use case. 

### Transactions
time | account | amount spent | overall balance after transaction
---- | ---------------- | ------ | -------------------------
2015-10-24 05:50:03 | A | 10 | 10
2015-10-24 05:51:18 | A |  5 |  15
2015-10-24 05:53:25 | B |  3 |  18
2015-10-24 05:57:11 | A | 17 | 35
2015-10-24 05:57:28 | A |  2 | 37

### Aggregated View
The aggregated view should look like demonstrated in the below table. 
The time column shows max(time) of each bucket, so I suppose it would make sense to tag the buckets with an aggregate function of the sorting criteria. Other use cases may require different aggregate functions for the sorting criteria (min, avg, ...). 

time | account | number of transactions | amount spent | overall balance after account transactions
---- | ---------------- | ---------------------- | ------ | -------------------------
2015-10-24 05:51:18 | A |  2 | 15 |  15
2015-10-24 05:53:25 | B |  1 | 3 |  18
2015-10-24 05:57:28 | A |  2 | 19 | 37

## More Complex Use Case

A more complex use case may require considering more than one discrimination criteria.
There is still only one sorting criteria. I'm not aware of a use case where more than one sorting criteria would make sense. 

### Transactions:

time | account | sub-account | amount spent | overall balance after transaction
---- | ------- | ----------- | ------------ | -------------------------
2015-10-24 05:50:03 | A | A1 | 10 | 10
2015-10-24 05:51:18 | A | A2 | 5 |  15
2015-10-24 05:53:25 | B | B1 | 3 |  18
2015-10-24 05:57:11 | A | A1 | 17 | 35
2015-10-24 05:57:28 | A | A1 | 2 | 37


### Aggregated View:

time | account | sub-account | number of transactions | amount spent | overall balance after account transactions
---- | ---------------- | ----------- | ---------------------- | ------ | -------------------------
2015-10-24 05:50:03 | A |  A1 | 1 | 10 |  10
2015-10-24 05:51:18 | A |  A2 | 1 |  5 |  15
2015-10-24 05:53:25 | B |  B1 | 1 | 3 |  18
2015-10-24 05:57:28 | A |  A1 | 2 | 19 | 37

**Comment by yaccob** [comment]
+1

**Comment by HonzaKral** [comment]
> I don't think it makes sense to implement this as a pipeline. Pipeline aggs only operate on the results of buckets produced by other aggs, which means all the documents have long since been lost to the query. So you'd be finding islands/gaps in bucketed data, which I don't think makes sense?

The idea is to use something like a (date) histogram with another nested aggregation (let's say `terms` with limit of 1 or `avg`) and then essentially grouping date intervals (days, months, minutes, ...) together based on continuous values of the inner aggregations (== days where XYZ is the most popular brand, hours when we are selling opposed to byuing, ....).

The problem is that people might be tempted to use overly fine aggregations (date_histogram over a year with a minute precision with a terms agg inside) to simulate the raw base aggregation. But that is a problem not specific to this use case.

**Comment by jariza** [comment]
+1

**Comment by castorm** [comment]
+1

**Comment by nik9000** [comment]
This is neat!

I see how doing this as a pipeline buys you sorting - and how it "automatically" solves the problem of merging the results from multiple shards. I guess the cost of doing it as a pipeline aggregation is that:

1. Its more confusing. @gingerwizard's use cases are all about documents. Doing it as a pipeline would actually have you detecting gaps in buckets.
2. It pushed lots of work to the coordinating node.

I bet we could make point 2 not so bad by doing stuff like streaming the results from the data nodes to the coordinating node rather than building them all at once (maybe we already do that). We could also push the calculations to the data node if we knew that all the data required was on the data node - say we used routing based on customer id or something.

I wonder about the temptation to use really really small aggregations under the pipeline - if you really do just want the documents maybe we can just do the search and read the data from doc_values somehow? That feels like it'd be much less "heavy". Still lots of data though.

I wonder if there are tricks we can play with statistics that let us push more stuff to the data nodes.

**Comment by bikeholik** [comment]
+1

**Comment by gschjetne** [comment]
This feature would be immensely useful for me, especially if one could specify bucketing criteria as a filter, ie. the aggregation returns buckets alternating between documents that match the filter and buckets that do not.

**Comment by polyfractal** [comment]
Been thinking about this.

This would be easiest done in pipeline aggregations.  If you embed a date_histo inside another date_histo, it would be reasonably simple to run through the buckets of the embedded histo to see when a list of tuples changes.  And then prune all buckets before the change.

It might technically be doable at a document level, but maybe not.  The islands could be collected locally, and overlapping buckets could be merged together during reduction like the Auto-Histo agg does (https://github.com/elastic/elasticsearch/pull/28993).  I haven't thought it through fully, but I suspect it might need to split buckets too.  

E.g. if one shard returns to islands which overlap another shard that has a single contiguous range... that contiguous range would need to be split.  Which may be impossible, since we've lost data at that point.

There's been a reasonable amount of support for this feature so I'm going to keep it up for now and chat with the team if we should prioritize it or not.

@elastic/es-search-aggs

**Comment by polyfractal** [comment]
We chatted about this today in the Search meeting.

First, we were curious if the new [Bucket Sort](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-pipeline-bucket-sort-aggregation.html) agg might help here?  It wasn't available at the time this ticket was opened.  It would provide a way to order sub-aggs so that the bucket-collapsing could be more easily done by the client.

Barring that, we identified three places this could potentially be implemented:

- A custom pipeline agg
- Extend bucket_selector to allow a more global view of buckets, so that it could select bucket on both sides of a "change" rather than just on the "current" bucket
- Extend bucket_script to allow removing buckets, so that this functionality could be handled purely in a script

We don't have a clear idea of how to best move forward at this point, but do see value in leaving this ticket open to see if there are any more comments.
